### PR TITLE
Added support for boolean matrix indexing (Fixes #2677)

### LINKED ISSUE #2677: Boolean indexing for SymPy Matrices
This should work for `a` as a SymPy Matrix:

```
a[[True, False, True]]
```

### PR DESCRIPTION
**Pull request by cpsquonk**

### CODE REVIEW

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -1035,6 +1041,10 @@ def extract(self, rowsList, colsList):
         """
         cols = self.cols
         flat_list = self._mat
+        if len(rowsList) > 0 and isinstance(rowsList[0], bool):
```
Comment (jcrist) [inline sympy/matrices/matrices.py]: For all cases of these, should be `if rowsList and isinstance(rowsList[0], bool)`...

Comment (jcrist) [inline sympy/matrices/matrices.py]: Actually, you should probably check if all elements are a boolean, not just the first one (just to be safe). Easy: `all(isinstance(i, bool) for i in rowsList)`.

File: sympy/matrices/tests/test_matrices.py [hunk 2]
```diff
@@ -72,6 +72,12 @@ def test_fancy_index_matrix():
         assert a[[0, 1], 2] == a[[0, 1], [2]]
         assert a[2, [0, 1]] == a[[2], [0, 1]]
         assert a[:, [0, 1]] == Matrix([[0, 1], [3, 4], [6, 7]])
+        assert a[[True, True, False], :] == Matrix([[0, 1, 2], [3, 4, 5]])
+        assert a[[True, True, False], 2] == a[[True, True, False],
+                                              [False, False, True]]
+        assert a[2, [True, True, False]] == a[[False, False, True],
+                                              [True, True, False]]
+        assert a[:, [True, True, False]] == Matrix([[0, 1], [3, 4], [6, 7]])
```
Comment (jcrist) [inline sympy/matrices/tests/test_matrices.py]: Needs tests for:

1. Sparse and Dense matrices
2. Indexing both rows and columns at the same time with boolean arrays

Also, in general, the right hand side of a test should be fixed, and not rely on the left hand side at all.

**Comment by jcrist** [comment]
Just a couple quick fixes, and this looks good. Since this has been a really long time since you submitted, if you don't want to finish this up let me know and I'll make the small fixes.

**Comment by czgdp1807** [comment]
Closing as per, https://github.com/sympy/sympy/issues/2677#issuecomment-558981761
Thanks @cap11235 for the contributions. Please let us know if you want to re-open this PR again.

**Pull request closed without merging**